File: stacked-ios/Stacked/Services/NoteRepository.swift
import Foundation
import Supabase

extension Notification.Name {
  static let notesDidChange = Notification.Name("notesDidChange")
}

@MainActor
final class NoteRepository {
  static let shared = NoteRepository()
  private var client: SupabaseClient { SupabaseService.client }
  private init() {}

  private func notifyChanged() {
    NotificationCenter.default.post(name: .notesDidChange, object: nil)
  }

  func fetchNotes() async throws -> [Note] {
    let rows: [NoteRowDTO] = try await client
      .from("notes")
      .select("id, title, body, color, pinned, created_at, updated_at")
      .is("archived_at", value: nil)
      .order("pinned", ascending: false)
      .order("updated_at", ascending: false)
      .execute()
      .value
    return rows.map(\.asNote)
  }

  func createNote(title: String?, body: String, color: NoteColor, pinned: Bool = false) async throws -> Note {
    guard let userId = client.auth.currentUser?.id else {
      throw NoteRepositoryError.notAuthenticated
    }
    let payload = NoteInsertDTO(
      user_id: userId,
      title: blankToNil(title),
      body: body,
      color: color.rawValue,
      pinned: pinned
    )
    let rows: [NoteRowDTO] = try await client
      .from("notes")
      .insert(payload)
      .select("id, title, body, color, pinned, created_at, updated_at")
      .execute()
      .value
    guard let note = rows.first?.asNote else {
      throw NoteRepositoryError.insertFailed
    }
    notifyChanged()
    return note
  }

  func updateNote(id: String, title: String?, body: String, color: NoteColor, pinned: Bool) async throws {
    let payload = NoteUpdateDTO(
      title: blankToNil(title),
      body: body,
      color: color.rawValue,
      pinned: pinned
    )
    try await client.from("notes").update(payload).eq("id", value: id).execute()
    notifyChanged()
  }

  func deleteNote(id: String) async throws {
    try await client.from("notes").delete().eq("id", value: id).execute()
    notifyChanged()
  }

  func setPinned(id: String, pinned: Bool) async throws {
    try await client.from("notes").update(["pinned": pinned]).eq("id", value: id).execute()
    notifyChanged()
  }

  private func blankToNil(_ value: String?) -> String? {
    let t = value?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return t.isEmpty ? nil : t
  }
}

enum NoteRepositoryError: LocalizedError {
  case notAuthenticated
  case insertFailed

  var errorDescription: String? {
    switch self {
    case .notAuthenticated: "Faça login para salvar notas."
    case .insertFailed: "Não foi possível criar a nota."
    }
  }
}

private struct NoteInsertDTO: Encodable {
  let user_id: UUID
  let title: String?
  let body: String
  let color: String
  let pinned: Bool
}

private struct NoteUpdateDTO: Encodable {
  let title: String?
  let body: String
  let color: String
  let pinned: Bool
}

private struct NoteRowDTO: Decodable {
  let id: String
  let title: String?
  let body: String
  let color: String
  let pinned: Bool
  let created_at: String?
  let updated_at: String?

  init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    if let s = try? c.decode(String.self, forKey: .id) { id = s }
    else { id = try c.decode(UUID.self, forKey: .id).uuidString }
    title = try c.decodeIfPresent(String.self, forKey: .title)
    body = try c.decodeIfPresent(String.self, forKey: .body) ?? ""
    color = try c.decodeIfPresent(String.self, forKey: .color) ?? NoteColor.mint.rawValue
    pinned = try c.decodeIfPresent(Bool.self, forKey: .pinned) ?? false
    created_at = try c.decodeIfPresent(String.self, forKey: .created_at)
    updated_at = try c.decodeIfPresent(String.self, forKey: .updated_at)
  }

  var asNote: Note {
    Note(
      id: id,
      title: title,
      body: body,
      color: NoteColor(rawValue: color) ?? .mint,
      pinned: pinned,
      createdAt: Self.parseTimestamp(created_at) ?? Date(),
      updatedAt: Self.parseTimestamp(updated_at) ?? Date()
    )
  }

  private static func parseTimestamp(_ raw: String?) -> Date? {
    guard let raw, !raw.isEmpty else { return nil }
    let withFraction = ISO8601DateFormatter()
    withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    let internet = ISO8601DateFormatter()
    internet.formatOptions = [.withInternetDateTime]
    return withFraction.date(from: raw) ?? internet.date(from: raw)
  }

  private enum CodingKeys: String, CodingKey {
    case id, title, body, color, pinned, created_at, updated_at
  }
}

File: stacked-web/lib/repositories/note-repository.ts
import { createClient } from "@/lib/supabase/client";

export type NoteColor = "mint" | "ash" | "amber" | "violet" | "rose";

export type Note = {
  id: string;
  title: string | null;
  body: string;
  color: NoteColor;
  pinned: boolean;
  createdAt: string;
  updatedAt: string;
};

type NoteRow = {
  id: string;
  title: string | null;
  body: string | null;
  color: string | null;
  pinned: boolean | null;
  created_at: string;
  updated_at: string;
};

const COLORS: NoteColor[] = ["mint", "ash", "amber", "violet", "rose"];

function mapRow(row: NoteRow): Note {
  const color = COLORS.includes(row.color as NoteColor) ? (row.color as NoteColor) : "mint";
  return {
    id: row.id,
    title: row.title,
    body: row.body ?? "",
    color,
    pinned: Boolean(row.pinned),
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function noteDisplayTitle(note: Note): string {
  const t = note.title?.trim() ?? "";
  if (t) return t;
  const first = note.body.split("\n")[0]?.trim() ?? "";
  return first || "Nota";
}

export function noteTaskSplit(note: Note): { title: string; description: string | null } {
  const t = note.title?.trim() ?? "";
  const body = note.body.trim();
  if (t) return { title: t, description: body || null };
  const lines = body.split("\n");
  const head = lines[0]?.trim() || "Nota";
  const rest = lines.slice(1).join("\n").trim();
  return { title: head, description: rest || null };
}

export class NoteRepository {
  static async list(): Promise<Note[]> {
    const supabase = createClient();
    const { data, error } = await supabase
      .from("notes")
      .select("id, title, body, color, pinned, created_at, updated_at")
      .is("archived_at", null)
      .order("pinned", { ascending: false })
      .order("updated_at", { ascending: false });
    if (error) throw error;
    return (data as NoteRow[] | null)?.map(mapRow) ?? [];
  }

  static async create(input: {
    title?: string | null;
    body: string;
    color: NoteColor;
    pinned?: boolean;
  }): Promise<Note> {
    const supabase = createClient();
    const {
      data: { user },
    } = await supabase.auth.getUser();
    if (!user) throw new Error("Faça login para salvar notas.");
    const { data, error } = await supabase
      .from("notes")
      .insert({
        user_id: user.id,
        title: input.title?.trim() || null,
        body: input.body,
        color: input.color,
        pinned: input.pinned ?? false,
      })
      .select("id, title, body, color, pinned, created_at, updated_at")
      .single();
    if (error) throw error;
    return mapRow(data as NoteRow);
  }

  static async update(
    id: string,
    input: { title?: string | null; body: string; color: NoteColor; pinned: boolean },
  ): Promise<void> {
    const supabase = createClient();
    const { error } = await supabase
      .from("notes")
      .update({
        title: input.title?.trim() || null,
        body: input.body,
        color: input.color,
        pinned: input.pinned,
      })
      .eq("id", id);
    if (error) throw error;
  }

  static async remove(id: string): Promise<void> {
    const supabase = createClient();
    const { error } = await supabase.from("notes").delete().eq("id", id);
    if (error) throw error;
  }

  static async convertToTask(note: Note): Promise<void> {
    const supabase = createClient();
    const {
      data: { user },
    } = await supabase.auth.getUser();
    if (!user) throw new Error("Faça login.");
    const { title, description } = noteTaskSplit(note);
    const { error: taskError } = await supabase.from("tasks").insert({
      titulo: title,
      descricao: description,
      user_id: user.id,
      concluida: false,
    });
    if (taskError) throw taskError;
    await this.remove(note.id);
  }
}
